### ISSUE TITLE
[v4.0.0-beta.8] Hardware back button not yet there

### ISSUE THREAD

**Issue by peterpeterparker** [ionic-team/ionic#15492]
# Feature Request

**Ionic Info**
Ionic Angular v4.0.0-beta.8

**Describe the Feature Request**
After having seen in the v4.0.0-beta8 CHANGELOG that the hardware back button was supported I had a try and unfortunately I noticed that it didn't worked out in case, at least, I would navigate using `navController.navigateForward`

In the other hand, I was able to close `Modal` using the back button successfully

**Describe Preferred Solution**
Being able to navigate back using the hardware back button while navigating with `navController.navigateForward`

**Related Code**
Sample repo: 

    https://github.com/peterpeterparker/ionic-back

**Additional Context**
I tested the issue on my Samsung Edge / Android v7

**Comment by ptitjes** [comment]
@peterpeterparker Yep, same here. With ionic beta.8/capacitor.beta7.

I did the following as a temporary hack (Note that `App.exitApp()` produces a strange android error message so I resolve to go back to my main `dashboard` view).

```ts

    // FIXME Hack because backButton is not supported correctly yet
    if (this.platform.is('android')) {
      fromEvent(App as NodeCompatibleEventEmitter, 'backButton').subscribe(async () => {
        try {
          let element = await this.actionSheetCtrl.getTop();
          if (element) {
            this.actionSheetCtrl.dismiss();
            return;
          }
        } catch (error) {
        }

        try {
          let element = await this.popoverCtrl.getTop();
          if (element) {
            this.popoverCtrl.dismiss();
            return;
          }
        } catch (error) {
        }

        try {
          let element = await this.modalCtrl.getTop();
          if (element) {
            this.modalCtrl.dismiss();
            return;
          }
        } catch (error) {
        }

        let canGoBack = this.routerOutlets.find((outlet: IonRouterOutlet) => outlet && outlet.canGoBack());
        if (canGoBack) {
          this.navCtrl.goBack();
          return;
        }

        this.navCtrl.navigateRoot('/dashboard');
      });
```

**Comment by seang96** [comment]
@ptitjes use navigator['app'].exitApp(); to exit the app.

**Comment by peterpeterparker** [comment]
I think it's solved in v4.0.0-beta.11

**Comment by Riyaz0001** [comment]
**Try This:** 
```

export class AppComponent {

    // set up hardware back button event.
    lastTimeBackPress = 0;
    timePeriodToExit = 2000;

    @ViewChildren(IonRouterOutlet) routerOutlets: QueryList<IonRouterOutlet>;

    constructor(
        private platform: Platform,
        private splashScreen: SplashScreen,
        private statusBar: StatusBar,
        public modalCtrl: ModalController,
        private actionSheetCtrl: ActionSheetController,
        private popoverCtrl: PopoverController,
        private router: Router,
        private toast: Toast) {
        this.initializeApp();
        // Initialize BackButton Eevent.
        this.backButtonEvent();
    }


// active hardware back button
    backButtonEvent() {

        // subscription to native back button
        this.platform.backButton.subscribe(async (data) => {
            alert(data);

            try {
                const element = await this.actionSheetCtrl.getTop();
                if (element) {
                    element.dismiss();
                    return;
                }
            } catch (error) {
            }

            try {
                const element = await this.popoverCtrl.getTop();
                if (element) {
                    element.dismiss();
                    return;
                }
            } catch (error) {
            }

            try {
                const element = await this.modalCtrl.getTop();
                if (element) {
                    element.dismiss();
                    return;
                }
            } catch (error) {
            }


            this.routerOutlets.forEach((outlet: IonRouterOutlet) => {
                if (outlet && outlet.canGoBack()) {
                    outlet.pop();
                } else if (this.router.url === '/home') {
                    // Double check to exit app
                    if (new Date().getTime() - this.lastTimeBackPress < this.timePeriodToExit) {
                        this.platform.exitApp(); // Exit from app

                    } else {
                        this.toast.show(
                            `Press back again to exit App.`,
                            '2000',
                            'center')
                            .subscribe(toast => {
                                // console.log(JSON.stringify(toast));
                            });
                        this.lastTimeBackPress = new Date().getTime();
                    }
                }
            });
        });
}
```
